nx/lib/nx.ex

```diff
@@ -3265,6 +3265,132 @@ defmodule Nx do
     end)
   end
 
+  @doc ~S"""
+  Split a tensor into train and test subsets.
+
+  `split` split must be either an integer greater than zero and less than the length of the tensor or a float number between `0.0` and `1.0`.
+
+  ## Options
+
+    * `:axis` - The axis along which to take the values from. Defaults to `0`.
+
+  ## Examples
+
+    All examples will operate on the same tensor so that it's easier to compare different configurations.
+
+    iex> t = Nx.tensor([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]])
+    iex> {left, right} = Nx.split(t, 2, axis: 0)
+    iex> left
+    #Nx.Tensor<
+      s64[2][4]
+      [
+        [0, 1, 2, 3],
+        [4, 5, 6, 7]
+      ]
+    >
+    iex> right
+    #Nx.Tensor<
+      s64[1][4]
+      [
+        [8, 9, 10, 11]
+      ]
+    >
+    iex> {left, right} = Nx.split(t, 2, axis: 1)
+    iex> left
+    #Nx.Tensor<
+      s64[3][2]
+      [
+        [0, 1],
+        [4, 5],
+        [8, 9]
+      ]
+    >
+    iex> right
+    #Nx.Tensor<
+      s64[3][2]
+      [
+        [2, 3],
+        [6, 7],
+        [10, 11]
+      ]
+    >
+    iex> {left, right} = Nx.split(t, 0.5, axis: 0)
+    iex> left
+    #Nx.Tensor<
+      s64[2][4]
+      [
+        [0, 1, 2, 3],
+        [4, 5, 6, 7]
+      ]
+    >
+    iex> right
+    #Nx.Tensor<
+      s64[1][4]
+      [
+        [8, 9, 10, 11]
+      ]
+    >
+    iex> {left, right} = Nx.split(t, 0.75, axis: 1)
+    iex> left
+    #Nx.Tensor<
+      s64[3][3]
+      [
+        [0, 1, 2],
+        [4, 5, 6],
+        [8, 9, 10]
+      ]
+    >
+    iex> right
+    #Nx.Tensor<
+      s64[3][1]
+      [
+        [3],
+        [7],
+        [11]
+      ]
+    >
+  """
+  @doc type: :indexed
+  def split(tensor, split, opts \\ [])
```

Comment:
@tiagodavi @josevalim I've pushed a refactor to collapse both clauses into the same one, mostly to highlight that the slicing is the same, with the only change being how we calculate it.

I've also changed the examples to operate on a written-out iota instead of arbitrary values because it makes it easier to compare results.

We still need to discuss if we want to accept a negative integer split.

It's easy to do it by setting right after `axis_size = `: `split = if is_integer(split) and split < 0, do: axis_size + split, else: split`

Reply:
Though I think that negative split might be confusing in the sense that we don't know if the results will be reversed or swapped in the result tuple (either or both)

Reply:
@polvalente we will base it on `Enum.split` (which means looking up from the end).

Reply:
But I am fine with postponing this for now. :)

Reply:
wdyt?